tensorflow/python/framework/tensor_util.py

```
@@ -252,7 +252,7 @@ def _FilterInt(v):
 def _FilterFloat(v):
   if isinstance(v, (list, tuple)):
     return _FirstNotNone([_FilterFloat(x) for x in v])
-  return None if isinstance(v, compat.real_types) else _NotNone(v)
+  return None if isinstance(v, compat.real_types) or v is None else _NotNone(v)
```

I don't understand this change: _NotNone(v) handles None in a specific way that seems intentional?  Can you clarify why you want to do this?

---

Looking at the original PR, I am going to defer to someone more of an expert on convert_to_tensor().  I don't know what it means to have a None value in an array constant.  It seems like this is targeting some kind of numpy parity but I don't know what numpy wants to do in the first place.

---

I'd defer further to @josh11b, since he wrote the original code that intentionally raises an exception (albeit perhaps not with the message intended) when `None` is present in a `tf.float32` or `tf.float64` constant. It looks like `tf.float16` slipped through the cracks when it was added, and we defer to NumPy semantics.

(My 2 cents: NumPy parity seems worth trying to achieve here, since I don't think it comes at a cost to backwards compatibility. However, I'm not sure what all the consequences of the change would be.)

---

Thanks @mrry, I'll defer to @josh11b then.  I don't know what it means to have a Tensor with some known values and some None values.

---

> Can you clarify why you want to do this?

The goal is better compatibility with numpy (this is described in the linked bug).

You can create a numpy float array from a list containing `None` and the `None` gets replaced by a `NaN`.

So the goal here is to accept `None` in the input data. 

Maybe this makes the call to `_NotNone()` redundant since, `_FilterFloat` returns `None` for success and the failing value on failure. So the purpose of `_NotNone` is to wrap `None` values so that failure can be reported for a `None` value.

I am no expert in this code either, this is just what I've understood from being involved in the bug.

---

> So the purpose of _NotNone is to wrap None values so that failure can be reported for a None value.

Right, so if the wrong data type is placed into a `tf.float32` tensor. The correct error message is returned.
```
sess = tf.Session()
print(sess.run(tf.constant([1, 2, 3, 4, 5, 6, 'foo'], dtype=tf.float32)))
# TypeError: Expected float32, got 'string' of type 'str' instead.
```

numpy floats allow `nan` values and numpy converts `None` to `nan`; however, the current `make_tensor_proto` and `tf.constant` do not convert `None` to `nan` for float types.

---

Right. I'm not sure if exact compatibility with NumPy was the goal with @josh11b added the `_NotNone()` path. It's possible that there was an overriding reason to exclude `None` values, which is why I'd like Josh to cast the deciding vote :).

---

I do not think silently converting None to NaN would be a good idea in general. At least from my personal experience the check to prevent Nones from being silently converted to tensors has saved me many hours of debugging weird issues.

---

@alextp, thanks for giving this a look over!

Currently there exists inconsistent behavior since `numpy.array` will convert `None` into `nan` for float types but, `convert_to_tensor` does not. This seems to break the promised `convert_to_tensor` abstraction referenced from issue #13827.

>Since the `convert_to_tensor` abstraction states the [following are equivalent](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/ops.py#L575-L578)
> ```
> value_1 = my_func(tf.constant([[1.0, 2.0], [3.0, 4.0]]))
> value_2 = my_func([[1.0, 2.0], [3.0, 4.0]])
> value_3 = my_func(np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32))
> ```
>I believe that the below should be equivalent as well
> ```
> value_a = my_func(tf.constant([[1.0, 2.0], [3.0, None]], dtype=tf.float32))
> value_b = my_func(np.array([[1.0, 2.0], [3.0, None]], dtype=np.float32))
> ```

From this inconsistency in `convert_to_tensor`, I believe there are a couple options.
1. We document that the abstraction does not hold for floats.
2. We remove the `convert_to_tensor` abstraction
3. We update `NumPy` to not convert `None` to `nan` for float types
4. We silently convert `None` to `nan` specifically for float types

Options 1 and 4 seem reasonable to me, but in my opinion option 4 is a good choice given that `convert_to_tensor` already silently converts and that it follows the semantics `NumPy` has set forth.

Please let me know what you think. Happy to make any needed changes!